### PR TITLE
MAINT: deal with cases in `minimize` for `(bounds.lb == bounds.ub).any()

### PR DESCRIPTION
**Pull request by andyfaff**
#### Reference issue
#12502

### CODE REVIEW

File: scipy/optimize/_differentiable_functions.py [hunk 1]
```diff
@@ -145,11 +145,17 @@ def update_grad():
                 self.g = grad_wrapped(self.x)
 
         elif grad in FD_METHODS:
+            lb, ub = _prepare_bounds(finite_diff_bounds, x0)
```
Comment (mdhaber) [inline scipy/optimize/_differentiable_functions.py:148]:
```
C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\site-packages\scipy\optimize\_numdiff.py:129: in _prepare_bounds
    lb = np.resize(lb, x0.shape)
E   AttributeError: 'list' object has no attribute 'shape'
```
Looks like `_prepare_bounds` doesn't like it if `x0` is a list, and apparently it's not guaranteed to be an array yet.

File: scipy/optimize/_numdiff.py [hunk 2]
```diff
@@ -491,32 +491,43 @@ def _dense_difference(fun, x0, f0, h, use_one_sided, method):
     h_vecs = np.diag(h)
 
     for i in range(h.size):
+        df = np.zeros_like(f0)
```
Comment (mdhaber) [inline scipy/optimize/_numdiff.py:494]: It looks like `df` will be all zeros if any of the `dx` are zero. Don't we want the elements of `df` corresponding with nonzero `dx` to be evaluated as usual?

File: scipy/optimize/_differentiable_functions.py (hunk 1 above)
Comment (andyfaff) [reply scipy/optimize/_differentiable_functions.py]: changed

File: scipy/optimize/_numdiff.py (hunk 2 above)
Comment (andyfaff) [reply scipy/optimize/_numdiff.py:494]: `dx` can either be scalar or `(N,)` length array.

If *all* of `dx` are zero then there don't need to be any evaluations of `fun(x)`; `df` should be 0 (or actually `np.zeros_like(f0)`), because the step is 0. This has the effect of reducing overall function evaluations.

If there are *any* non-zero `dx` then `fun(x)` is evaluated to calculate `df` as usual.

**Comment by andyfaff** [comment]
The last commit checks that `ScalarFunction` rejects functions that don't return scalars.

**Comment by andyfaff** [comment]
@mdhaber I think the failing tests are due to CI not liking the latest pytest.

File: scipy/optimize/_numdiff.py (hunk 2 above)
Comment (mdhaber) [reply scipy/optimize/_numdiff.py:494]: <face-palm> i was tired.

File: scipy/optimize/_differentiable_functions.py [hunk 3]
```diff
@@ -145,11 +150,17 @@ def update_grad():
                 self.g = grad_wrapped(self.x)
 
         elif grad in FD_METHODS:
+            lb, ub = _prepare_bounds(finite_diff_bounds, self.x)
```
Comment (pv) [inline scipy/optimize/_differentiable_functions.py:153]: Shouldn't this be inside `update_grad`?
`self.x` presumably is not a constant.

Comment (pv) [reply scipy/optimize/_differentiable_functions.py:153]: Never mind, `_prepare_bounds` only needs `x.shape`.

**Comment by andyfaff** [comment]
rebased to make tests run again.
@mdhaber, ping for merge?

**Comment by mdhaber** [comment]
I'm afraid I can't check this weekend : / I'm swamped. Is in time for 1.6.0 OK?
Tell you what, I'll trade you for taking a look at gh-12043 : P  (kidding)

**Comment by andyfaff** [comment]
> I'm afraid I can't check this weekend : / I'm swamped. Is in time for 1.6.0 OK?
> Tell you what, I'll trade you for taking a look at [gh-12043](https://github.com/scipy/scipy/pull/12043) : P (kidding)

1.6.0 is fine, I was just trying to clear low hanging fruit. In all seriousness I would review your PR, but I'm not an expert in the subject matter.

**Comment by andyfaff** [comment]
@mdhaber I was fiddling around with the idea of factorising out equal bounds (reducing dimensionality of the problem), but it's not only the objective function one has to worry about. One also have to deal with jacobians and hessians, and I'm not sure if it's possibly to do that in a clean way.

**Comment by mdhaber** [comment]
User provided functions for those, right? Why can't we eliminate the rows and columns of those outputs that correspond with the fixed variables?

**Comment by andyfaff** [comment]
Ahhh, filter afterwards rather than before.

**Comment by mdhaber** [comment]
Yeah, there's nothing we can do to stop them from being evaluated, but we can remove them.

I am imagining that `minimize` identifies equal bounds, removes those variables from the decision variable vector before calling the underlying solver. To the solver it passes not the user provided functions but wrappers that inject the fixed variables into the input and remove the derivatives corresponding with the fixed variables from the output.

File: scipy/optimize/tests/test_optimize.py [hunk 4]
```diff
@@ -2174,8 +2176,37 @@ def test_memoize_jac_with_bfgs(function_with_gradient):
...
+        # warning filter is for trust-constr
+        # UserWarning: delta_grad == 0.0.Check if the approximated
+        # function is linear.
+        warnings.simplefilter('ignore', category=UserWarning)
+
+        res = optimize.minimize(
+            optimize.rosen, x0, method=method, bounds=bounds,
+        )
+        assert res.success
+        assert_allclose(res.x, np.r_[best_x.x, 2.0], rtol=3e-6)
```
Comment (larsoner) [inline scipy/optimize/tests/test_optimize.py:2212]: I assume a more modern (and compact) way to do this would be:
```suggestion
    with pytest.warns(UserWarning, match='.*Check if the.*is linear.*'):
        res = optimize.minimize(
            optimize.rosen, x0, method=method, bounds=bounds,
        )
    assert res.success
    assert_allclose(res.x, np.r_[best_x.x, 2.0], rtol=3e-6)
```

File: scipy/optimize/tests/test_differentiable_functions.py [hunk 5]
```diff
@@ -326,6 +326,13 @@ def f(x):
         assert_equal(f2, 14.0)
         assert x is not sf.x
 
+    def test_non_scalar_func(self):
+        # Use of a non-scalar function should raise an error when using
+        # ScalarFunction
+        with assert_raises(ValueError):
```
Comment (larsoner) [inline scipy/optimize/tests/test_differentiable_functions.py:332]: The newer `pytest` syntax is a bit nicer IMO:
```suggestion
        with pytest.raises(ValueError, match='must return a scalar):
```

**Review by larsoner — APPROVED** [review]
Just a couple of minor testing ideas -- +1 for merge with or without these suggestions

File: scipy/optimize/_differentiable_functions.py [hunk 6]
```diff
@@ -145,11 +151,17 @@ def update_grad():
...
+            equal_bounds = lb == ub
+
             def update_grad():
                 self._update_fun()
                 self.ngev += 1
                 self.g = approx_derivative(fun_wrapped, self.x, f0=self.f,
                                            **finite_diff_options)
+                # when the lower and upper bounds are equal then set the
+                # gradient to 0.
+                self.g[equal_bounds] = 0.0
```
Comment (kif) [inline scipy/optimize/_differentiable_functions.py:164]: This is the core of the PR: set to 0 derivatives of fixed parameters (which were Nans). I wish it is back-ported to the 1.5 series.

**Comment by kif** [comment]
For me, the main reason to use this "feature" (bounds.ub == bounds.lb) to fix a parameter is because re-rewriting the function (and its derivative) is probably more complicated & error prone.

This PR basically enforces the derivative of the function along fixed parameters to be zero instead of 0/0=Nan. 

Beside this, looks good to me.

File: scipy/optimize/tests/test_differentiable_functions.py (hunk 5 above)
Comment (andyfaff) [reply scipy/optimize/tests/test_differentiable_functions.py:332]: Ack, changed.

File: scipy/optimize/tests/test_optimize.py (hunk 4 above)
Comment (andyfaff) [reply scipy/optimize/tests/test_optimize.py:2212]: Thanks for the suggestion here. However only trust-constr issues that warning, not all the minimize methods. So I think it's better to filter the warning for trust-constr.

**Comment by andyfaff** [comment]
Test failures are unrelated.

**Comment by andyfaff** [comment]
@larsoner rebased to fix merge conflict.

**Comment by andyfaff** [comment]
@larsoner ping for a merge?

File: scipy/optimize/_numdiff.py [hunk 7]
```diff
@@ -551,32 +551,43 @@ def _dense_difference(fun, x0, f0, h, use_one_sided, method):
...
+
         if method == '2-point':
             x = x0 + h_vecs[i]
             dx = x[i] - x0[i]  # Recompute dx as exactly representable number.
-            df = fun(x) - f0
+            if np.count_nonzero(dx):
```
Comment (nmayorov) [inline scipy/optimize/_numdiff.py:559]: For my understanding: `dx` is a float here, why not simply write `dx != 0`?

File: scipy/optimize/_numdiff.py [hunk 8]
```diff
@@ -551,32 +551,43 @@ def _dense_difference(fun, x0, f0, h, use_one_sided, method):
...
             dx = h_vecs[i, i]
         else:
             raise RuntimeError("Never be here.")
 
-        J_transposed[i] = df / dx
+        with np.errstate(invalid='ignore'):
+            # filter warning for 0 / 0
+            # However a 1.0 / 0.0 will still raise a
+            # RuntimeWarning: divide by zero encountered in true_divide
+            g = np.asarray(df / dx)
```
Comment (nmayorov) [inline scipy/optimize/_numdiff.py:588]: What will be result in `g` when `df` and `dx` are zeros? I thought `nan`, but apparently is `0`. 

Should it be documented, that in case `lb[i] == ub[i]` the corresponding column in Jacobian will be zero?

File: scipy/optimize/_differentiable_functions.py (hunk 6 above)
Comment (nmayorov) [reply scipy/optimize/_differentiable_functions.py:164]: It seems that with the change in `approx_derivative`, `self.g[equal_bounds]` will be filled with zeros by `approx_derivative`. Filling it explicitly is fine too.

Comment (nmayorov) [reply scipy/optimize/_differentiable_functions.py:164]: A more important question: could someone explain qualitatively why this change is fundamentally correct?

I don't want to crash into this PR all of the sudden and destroy all the work, but I have some concerns.

1. From a logical/mathematical point of view, the fact that `lb[i] == ub[i]` doesn't impose any constraints on the gradient at `lb[i]`. So it's very unclear what setting elements of `g` to zero achieves.
2. What is going to happen if we run the same problem now with analytical and numerical differentiation and some bounds being equal? In principle we must get close and correct results (including `x_opt[i] == lb[i] == ub[i]`), but I don't see how it is guaranteed with the current code.
3. It seems to me that the ability of a solver to handle `lb[i] == ub[i]` correctly with all possible settings requires more considerations and testing.

File: scipy/optimize/_numdiff.py (hunk 7 above)
Comment (andyfaff) [reply scipy/optimize/_numdiff.py:559]: `dx` can also be an array

File: scipy/optimize/_numdiff.py (hunk 8 above)
Comment (andyfaff) [reply scipy/optimize/_numdiff.py:588]: You're right that in floating point arithmetic if `df` and `dx` are zeros then `df/dx` is `nan`. If people set `lb[i] == ub[i]` from a bounded optimizer that calculates gradients then this calculation will emit a RuntimeWarning:

```
RuntimeWarning: invalid value encountered in true_divide
  np.array([0.0]) / np.array([0.0])
```

File: scipy/optimize/_differentiable_functions.py (hunk 6 above)
Comment (andyfaff) [reply scipy/optimize/_differentiable_functions.py:164]: I'm less concerned about losing work for the PR, and more concerned that we get behaviour right here. It'd be great to have more input with people familiar with optimisation, to make sure that happens. @nmayorov I'd appreciate any further input you're able to give.

The changes in this PR are by no means perfect.

1. Logically setting `lb[i] == ub[i]` means that you know, with 100% certainty, that `x[i]` has to be a constant value. Should a function have a gradient for `x[i]` if it's not allowed to take any other value? A graph of f vs x would show a singular dot, and not a line. In such a case the gradient should probably be undefined (nan), even for an 'analytic' derivative (or should it be zero?).

2. An analytic jacobian function (derived for `lb[i] < ub[i]`) may be able to calculate the gradient at `x[i] == lb[i]` with `lb[i] == ub[i]`. However, estimating the gradient with finite differences at that point *and respecting bounds* is impossible. If bounds are ignored during numeric differentiation, then it is possible to do that (you're also saying that you're not 100% sure that `lb[i] == ub[i]`). 
There were two sets of issues reported that have been involved in the lead to this PR:

- (a) estimating grad via finite differences should respect bounds. This lead to the use of `approx_derivative` that does so.
- (b) when the bounds-respecting was done for the previous issue this lead to issues in numerical optimizers, such as `slsqp` or `lbfgsb` that don't work when the jacobian contains `nan` (`df/dx=0.0/0.0`)

From testing SLSQP and LBFGSB do work when those nan are replaced by zeros, so that was the approach I took here. A different approach is to recognise problems with equal bounds and factor out those fixed parameters, thereby reducing the dimensionality of the problem. I think this factorisation would increase code complication and not straightforward to implement across all the solvers.
Yet another approach could be to get the actual iterative code in SLSQP and LBFGSB to behave properly for `lb[i] == ub[i]`, I don't think anyone has the appetite for altering Fortran code.

I don't think that this PR, or an alternate that factored out fixed parameters, would lead to the analytic or numerical differentation approaches giving the same answer unless we rolled back the changes from 2a.

TLDR: even though it's not a great solution, it may be the best of a bad lot.

Comment (andyfaff) [reply scipy/optimize/_differentiable_functions.py:164]: If one of the stumbling blocks is the alteration of nan-->0 in `_differentiable_functions`, then that could be done in a wrapper function at a higher level.

File: scipy/optimize/_numdiff.py (hunk 8 above)
Comment (nmayorov) [reply scipy/optimize/_numdiff.py:588]: I saw `np.array(0) / np.array(0) = np.array(0)` (with a warning) in numpy 1.13 and Python 2.7.

I think `nan` is an expected result now (as it should be in my opinion). Not sure if it worth investigating in detail.

File: scipy/optimize/_numdiff.py (hunk 7 above)
Comment (nmayorov) [reply scipy/optimize/_numdiff.py:559]: Not for the sake of proving you wrong, but after all ifs `df / dx` is computed, which corresponds to a column in Jacobian. And just by indexing `dx = x[i] - x0[i]` it really looks like a float. Maybe I'm missing something.

It's not terribly important, but if it is a float, then maybe it's better not to use `np.count_nonzero`.

Comment (andyfaff) [reply scipy/optimize/_numdiff.py:559]: `h_vecs` can be a array size > 1.

Comment (nmayorov) [reply scipy/optimize/_numdiff.py:559]: Somewhat tricky logic is involved here.

`h_vecs = np.diag(h)` and `x = x0 + h_vecs[i]` - meaning that we add deltas of the form `[0, 0, h[2], 0, 0, 0]`, etc. And after that we compute scalar `dx = x[i] - x0[i]`, which will be very close to `h[i]` but more accurate considering floating-point arithmetic (the comment `# Recompute dx as exactly representable number.`).

At the end `J_transposed[i] = df / dx` -- i-th column of the Jacobian is vector `df` divided by scalar `dx`.

File: scipy/optimize/_numdiff.py (hunk 8 above)
Comment (andyfaff) [reply scipy/optimize/_numdiff.py:588]: I agree with you in principle that `approx_derivative` should return nan for equal bounds. But if `approx_derivative` returns a `nan` for those entries, and emits a RuntimeWarning, then the calling code in various `minimize` methods will need to:

- filter the RuntimeWarning, because users wouldn't expect a warning from setting `lb[i] == ub[i]`, which has historically been allowed.
- change all the nan entries in the gradient to 0 because the SLSQP/LBFGSB/etc Fortran code can't deal with nan entries in a grad matrix.

Overall we need to come up with a fix for #12502. There are several other duplicate issues, so it's a pain point.

File: scipy/optimize/_numdiff.py (hunk 7 above)
Comment (andyfaff) [reply scipy/optimize/_numdiff.py:559]: Using `count_nonzero` means that we can avoid function evaluations for which we know all entries in `dx` are zero. I should've said in my previous comment that `dx` can have size > 1, it's not necessarily a scalar.

Comment (andyfaff) [reply scipy/optimize/_numdiff.py:559]: Using `np.count_nonzero` or `if dx != 0` allows one to avoid a function evaluation if the step size is zero, which it will be for equal bounds.

File: scipy/optimize/_differentiable_functions.py (hunk 6 above)
Comment (nmayorov) [reply scipy/optimize/_differentiable_functions.py:164]: I understand that you want to fix a problem with optimization with equal bounds failing due to valid restrictions of numerical differentiation.

I think we can split the problem into 2 parts:

1. Determine which solvers support equal bounds correctly. If the solver doesn't support it -- we should raise an error immediately in all cases.
2. If the solver does support equal bounds and the problem is formulated with equal bounds and numerical differentiation is requested then we have 3 options: raise an appropriate error, remove restrictions on allowed steps in `approx_derivative` or try the trick with zero gradient proposed here. 

I'll try to think some more about the situation and investigate how existing solvers supposed to work with fixed variables.

----

And I actually like the idea of automatic pre-elimination of fixed variables, it seems like a universal and correct approach. Not impossible to implement too.

Comment (andyfaff) [reply scipy/optimize/_differentiable_functions.py:164]: Elimination has a few issues. User functions have to be sent the full array, but the solvers have to be provided with a reduced set. So one has to write a wrapper for the actual problem, but also wrappers for any constraints machinery (e.g. slsqp). Finally one has to expand what the (fortran) solver returns so that OptimizeResult contains the correct number of variables.
i think there is large amount of tinkering required for each minimizer method, I don't think it's possible at a higher level.

**Comment by andyfaff** [comment]
@nmayorov @larsoner, I've made an alternate version, https://github.com/andyfaff/scipy/pull/21. In that version the replacement of nan-->0 occurs within a wrapper in `_minimize_*`, and not in `approx_derivative`. @nmayorov, would this be preferred to this PR?

File: scipy/optimize/_differentiable_functions.py (hunk 6 above)
Comment (nmayorov) [reply scipy/optimize/_differentiable_functions.py:164]: Not 100% sure, but conceptually it looks feasible using Python-side wrappers. The problem must be completely reformulated  to remove fixed variable, then solved by the optimizer, then the result is restored to contain the fixed variables.

Comment (andyfaff) [reply scipy/optimize/_differentiable_functions.py:164]: @nmayorov, the elimination solution is likely to take some time, and I think it's vital to get at least a temporary fix in for the next release. At the moment this is either going to be this PR, or #13069. The alternate PR lifts the issue from the numdiff code up into the solver method code.

File: scipy/optimize/_numdiff.py (hunk 7 above)
Comment (nmayorov) [reply scipy/optimize/_numdiff.py:559]: Yeah, I see. But `np.count_nonzero` seems out of place a little bit, considering that `dx` is float (rechecked by adding additional `assert np.array(dx).ndim == 0` and running all the tests).

File: scipy/optimize/_differentiable_functions.py (hunk 6 above)
Comment (nmayorov) [reply scipy/optimize/_differentiable_functions.py:164]: Are you sure it's vital to get such an ad-hoc fix? I mean if it fails with an error - it's an expected behavior and the error message is correct in this case. The case with equal bounds seems to be  "undefined area" in the current state to be honest. 

Well I'm not actively involved in the development as of late, so don't want to block anything. The second PR looks preferable to me.

As for `_approx_derivative` - maybe it should even raise an error with equal bounds, because it's incorrect setting considering the logic involved. But it won't play well with you changes, so it just a thought.

Comment (andyfaff) [reply scipy/optimize/_differentiable_functions.py:164]: The reported issues (+ their volume) and the [documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.Bounds.html#scipy.optimize.Bounds) for minimize indicate that setting equal bounds has historically been acceptable/valid.
However, the last release made the numerical differentiation obey any bounds, so back compatibility with this historical behaviour was broken (because the core of slsqp+lbfgsb can't cope with NaN in the gradient). Squash one bug, and another pops up.
Given that the alternate approach in #13069 might be preferable, would you have time to review that PR? It would be so, so, helpful. I agree that the change of changing Nan entries to zero in the gradient isn't 100% desirable. However, for the test example I constructed the modifications worked, they also work for various users posting these issues.

Comment (mdhaber) [reply scipy/optimize/_differentiable_functions.py:164]: @andyfaff @nmayorov do you see problems with the approach in gh-13096 (different from gh-13069)?

**Comment by tylerjereddy** [comment]
milestone bumped, if this really is important for the 1.6.x series you may still have a tiny bit of time, but I'm trying to branch soon

File: scipy/optimize/_numdiff.py (hunk 7 above)
Comment (andyfaff) [reply scipy/optimize/_numdiff.py:559]: I agree with your arguments @nmayorov, made those changes.

File: scipy/optimize/_numdiff.py (hunk 8 above)
Comment (andyfaff) [reply scipy/optimize/_numdiff.py:588]: @nmayorov, these changes to `approx_derivative` will return `np.nan` if a lower bound is equal to an upper bound. I've noted that in the docstring.
Note: those np.nans are then converted to zero in `ScalarFunction`.

**Comment by andyfaff** [comment]
@mdhaber, @nmayorov, @larsoner on the whole I prefer this option rather than Matt's alternative in #13096, because it seems simpler and more maintainable.

Merging this PR would recover the ability to have lower bounds equal to upper bounds for several solvers (a long standing issue for L-BFGS-B, SLSQP).
Remaining issues for discussion:

- the behaviour for analytical and finite-difference gradients would be different. This PR sets the gradient entries for equal bounds to 0. Corresponding entries for analytic gradients would be different. For me the a gradient of zero for a fixed parameter seems logical. When using these kinds of minimisers I'd certainly expect an inverse Hessian (c.f. covariance matrix) to have zero entries for those parameters, because they're fixed.

File: scipy/optimize/_numdiff.py (hunk 2 above)
Comment (mdhaber) [reply scipy/optimize/_numdiff.py:556]: I wonder if the tests are hitting the array `dx` case. I think you will get a:
```
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

How about `np.any(dx)`?

File: scipy/optimize/tests/test_optimize.py [hunk 9]
```diff
@@ -2308,3 +2309,31 @@ def fquad_jac(x):
             hess = fquad_hess
         res = optimize.minimize(fquad, x0, method=meth, jac=jac, hess=hess)
         assert_allclose(res.x, np.arange(np.size(x0)), atol=2e-4)
+
+
+@pytest.mark.parametrize('method', ['Powell', 'L-BFGS-B', 'SLSQP',
+                                    'trust-constr'])
```
Comment (mdhaber) [inline scipy/optimize/tests/test_optimize.py:2315]: I think we'll need more comprehensive tests, as this currently passes in master for Powell and trust-constr.

Comment (andyfaff) [reply scipy/optimize/tests/test_optimize.py:2315]: I can't remember from the original issue if Powell and trust-constr were sensitive to this. I can only remember that lbfgsb and slsqp were.

File: scipy/optimize/_numdiff.py (hunk 2 above)
Comment (andyfaff) [reply scipy/optimize/_numdiff.py:556]: for future reference `dx` is an array of size > 1 when evaluating a hessian.
